app/services/container_dashboard_service.rb

```diff
@@ -228,6 +226,64 @@ def daily_network_metrics
     end
   end
 
+  def beginning_of_hour(timestamp)
+    timestamp.beginning_of_hour.utc
+  end
+
+  def beginning_of_day(timestamp)
+    timestamp.strftime("%Y-%m-%d")
+  end
+
+  def fill_pod_metrics(metrics, timestamp_func, pods_create_trend, pods_delete_trend)
+    metrics.each do |m|
+      timestamp = send(timestamp_func, m.timestamp)
+
+      pods_create_trend[timestamp] += m.stat_containergroup_create_rate.to_i
+      pods_delete_trend[timestamp] += m.stat_containergroup_delete_rate.to_i
+    end
+  end
+
+  def hourly_pods_metrics
+    hourly_pods_create_trend = Hash.new(0)
+    hourly_pods_delete_trend = Hash.new(0)
+    metrics = hourly_metrics
+
+    fill_pod_metrics(metrics, "beginning_of_hour",
+                     hourly_pods_create_trend, hourly_pods_delete_trend)
+
+    if metrics.any?
+      {
+        :xData  => hourly_pods_create_trend.keys,
+        :yData  => hourly_pods_create_trend.values.map(&:round),
+        :yData1 => hourly_pods_delete_trend.values.map(&:round)
+      }
+    end
+  end
+
+  def daily_pods_metrics
+    daily_pods_create_trend = Hash.new(0)
+    daily_pods_delete_trend = Hash.new(0)
+
+    fill_pod_metrics(daily_provider_metrics, "beginning_of_day",
+                     daily_pods_create_trend, daily_pods_delete_trend)
+
+    if daily_provider_metrics.any?
+      {
+        :xData  => daily_pods_create_trend.keys,
+        :yData  => daily_pods_create_trend.values.map(&:round),
+        :yData1 => daily_pods_delete_trend.values.map(&:round)
+      }
+    end
```

Comment:
The 2 functions hourly/daily_pods_metrics() are still almost identical.
fill_pod_metrics() would have a cleaner interface if it didn't take hashes as "output parameters" but created them itself (it might need to be renamed then).
The xData, yData, yData1 hash creation could also be factored out but probably doesnt belong in fill_pod_metrics().

daily/hourly_network_metrics() are also very similar.
How about a single `bin_metrics()` function summing just one attribute?
Then you can write something like:

```
  def hourly_network_metrics
    hourly_net = bin_metrics(:beginning_of_hour, &:net_usage_rate_average)
    if daily_provider_metrics.any?
      {
        :xData => hourly_net.keys,
        :yData => hourly_net.values.map(&:round)
      }
    end
  end

  def daily_pods_metrics
    if daily_provider_metrics.any?
      create_trend = bin_metrics(:beginning_of_day, { |m| m.stat_containergroup_create_rate.to_i })
      delete_trend = bin_metrics(:beginning_of_day, { |m| m.stat_containergroup_delete_rate.to_i })
      {
        :xData  => create_trend.keys
        :yData  => create_trend.values.map(&:round),
        :yData1 => delete_trend.values.map(&:round)
      }
    end
  end
````
and similarly for the other functions...  Even ems_utilization() could use this.

Subtle point: you're relying on order of keys and values across 2 separate hashes.
That works because they're both filled in same order from iterating on the same input.
It becomes even subtler with my proposal when they're created in 2 separate calls.
Not sure if that's fine, worth a comment or justifies another approach (I'm still not used to ruby hashes being ordered...)